# Mid Level Descriptors term paper (Notebook 1/2)

by Marie Orschulik (Matrikel number: 411093)

In [1]:
#import sys,os
#sys.path.append(os.getcwd()+"/my_mld")

import sys
sys.path.append('./mld/src')
import os
import pandas as pd
import numpy as np
import IPython
import opensmile
sys.path.append("./home/audeering.local/morschulik/S_SprechStile_Indv/text_features/")
#import nltk
sys.path.append("./text_features/")
import text_features as tf

In [2]:
import opensmile

In [3]:
import midlevel_descriptors as mld

In [4]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to
[nltk_data]     /home/audeering.local/morschulik/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/audeering.local/morschulik/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

#### Load the audio file in form of a Python object

In [5]:
df_q = pd.read_pickle('processed/Orschulik/Orschulik_mdf.pkl')
df_q.head(1)

channel  \
file                                 start                  end                               
./tmp/speeches/qstimme_Orschulik.wav 0 days 00:00:04.240000 0 days 00:00:06.620000        0   

                                                                                                                       segmented_file  
file                                 start                  end                                                                        
./tmp/speeches/qstimme_Orschulik.wav 0 days 00:00:04.240000 0 days 00:00:06.620000  /home/felix/data/TU/SS2021/S_Sprechstil_Andere...

### Rearrange the dataframe
Here we create a function to rearrange our pandas dataframe to a dataframe with "segemted_file" as index.
Inside the function we dropped all colums, but the "segmented_file"-column and set it as index.
All the segmented files are inside the new folder "all_wavs"  that is set as a new audio directory.

In [6]:
df_q = pd.read_pickle('./processed/Orschulik/Orschulik_mdf.pkl')
def re_arrange_df(df, audio_dir):
    df['segmented_file'] = df['segmented_file'].apply(lambda x: audio_dir+os.path.basename(x))
    df = df.set_index(df.segmented_file)
    df = df.drop(columns=['segmented_file', 'channel'])
    return df
df_q = re_arrange_df(df_q, 'all_wavs/')

In [7]:
df_q.head(1)

Empty DataFrame
Columns: []
Index: [all_wavs/qstimme_Orschulik_seg_00000000.wav]

In [8]:
IPython.display.Audio('Speeches/qstimme_Orschulik.wav')

In [9]:
IPython.display.Audio(df_q.index[0])

### mid-level-descriptor (MLD) extraction

In [10]:
fex_mld = mld.MLD()
mld_df = fex_mld.extract_from_index(index=df_q, cache_path='./tmp/mld')

### We set up a feature extractor for low level descriptors and functionals of a pre-defined feature set

In [11]:
smile_lld = opensmile.Smile(
    feature_set = opensmile.FeatureSet.GeMAPSv01b,
    feature_level = opensmile.FeatureLevel.LowLevelDescriptors,
)
smile_functionals = opensmile.Smile(
    feature_set = opensmile.FeatureSet.GeMAPSv01b,
    feature_level = opensmile.FeatureLevel.Functionals,
)

In [12]:
lld_df = smile_lld.process_files(df_q.index)
print(lld_df.shape)

/home/audeering.local/morschulik/S_SprechStile_Indv/my_mld/lib/python3.8/site-packages/audinterface/core/process.py:127: FutureWarning: In a future version of pandas all arguments of MultiIndex.set_levels except for the argument 'levels' will be keyword-only
  y.index = y.index.set_levels([[start], [end]], [1, 2])


(3526, 18)


In [13]:
funct_df = smile_functionals.process_files(df_q.index)
print(funct_df.shape)

(13, 62)


In [14]:
lld_df.head(3)

Loudness_sma3  \
file                                        start                  end                                     
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000       0.480453   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000       0.637537   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000       0.778748   

                                                                                           alphaRatio_sma3  \
file                                        start                  end                                       
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000        -8.431729   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000       -12.139736   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000       -15.712261   

                                                                                           hammarbergIndex_sma3  \
file                                        start                  end                                            
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000             18.233580   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000             23.054647   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000             29.612722   

                                                                                           slope0-500_sma3  \
file                                        start                  end                                       
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000         0.024905   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000         0.011771   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000         0.014289   

                                                                                           slope500-1500_sma3  \
file                                        start                  end                                          
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000           -0.006906   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000            0.001506   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000            0.009733   

                                                                                           F0semitoneFrom27.5Hz_sma3nz  \
file                                        start                  end                                                   
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000                    27.435410   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000                    27.388617   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000                    27.282709   

                                                                                           jitterLocal_sma3nz  \
file                                        start                  end                                          
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 00:00:00        0 days 00:00:00.020000                 0.0   
                                            0 days 00:00:00.010000 0 days 00:00:00.030000                 0.0   
                                            0 days 00:00:00.020000 0 days 00:00:00.040000                 0.0   

                                                                                           shimmerLocaldB_sma3nz  \
file         

The low level descriptors describe each 10 ms of the original audio file. The low level descriptors dataframe has the dimensionality: 3526 rows and 18 columns.

In [15]:
funct_df.head(3)

F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                        start  end                                                         
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                          28.023560   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                          26.052574   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                          26.504032   

                                                                           F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                        start  end                                                              
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                0.098374   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                0.097217   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                                0.164797   

                                                                           F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                        start  end                                                                  
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                   26.182751   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                   24.033720   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                                   24.153654   

                                                                           F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                        start  end                                                                  
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                   28.269264   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                   25.690302   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                                   25.857399   

                                                                           F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                        start  end                                                                  
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                   29.687143   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                   28.134933   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                                   28.203560   

                                                                           F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                        start  end                                                                
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                  3.504393   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                  4.101213   
all_wavs/qstimme_Orschulik_seg_00000002.wav 0 days 0 days 00:00:03.720000                                  4.049906   

                                                                           F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                        start  end                                                                   
all_wavs/qstimme_Orschulik_seg_00000000.wav 0 days 0 days 00:00:02.380000                                    24.140854   
all_wavs/qstimme_Orschulik_seg_00000001.wav 0 days 0 days 00:00:03.580000                                    80.314476   
all_wavs/qstimme_Orschulik

The functionals dataframe consists of 13 rows because of 13 segmented_files in which the original audio file was segmented. Each segment is described by a vector of fixed dimensionality after applying the following functionals:

* Extreme values and positions

* Means (arithmetic, quadratic, geometric)

* Moments (standard deviation, variance, kurtosis, skewness)

* Percentiles and percentile ranges

* Regression (linear and quadratic approximation, regression error)

* Centroid

* Peaks

* Segments

* Sample values

* Times/durations

* Onsets/Offsets

* Discrete Cosine Transformation (DCT)

* Zero Crossings

* Linear Predictive Coding (LPC) coefficients and gain


The functionals dataframe has the dimensionality: 13 rows and 62 columns. 

In [16]:
mld_df.head(3)

hld_sylRate  hld_gapRel  \
file                                                                   
all_wavs/qstimme_Orschulik_seg_00000000.wav     3.652968    0.079808   
all_wavs/qstimme_Orschulik_seg_00000001.wav     4.204204    0.069816   
all_wavs/qstimme_Orschulik_seg_00000002.wav     4.761905    0.096759   

                                             hld_nSyl  \
file                                                    
all_wavs/qstimme_Orschulik_seg_00000000.wav         8   
all_wavs/qstimme_Orschulik_seg_00000001.wav        14   
all_wavs/qstimme_Orschulik_seg_00000002.wav        16   

                                             lld_voicingFinalUnclipped_sma_diff  \
file                                                                              
all_wavs/qstimme_Orschulik_seg_00000000.wav                            0.066320   
all_wavs/qstimme_Orschulik_seg_00000001.wav                           -0.059167   
all_wavs/qstimme_Orschulik_seg_00000002.wav                           -0.024086   

                                             lld_voicingFinalUnclipped_sma_end  \
file                                                                             
all_wavs/qstimme_Orschulik_seg_00000000.wav                          -0.035815   
all_wavs/qstimme_Orschulik_seg_00000001.wav                           0.001708   
all_wavs/qstimme_Orschulik_seg_00000002.wav                          -0.140285   

                                             lld_voicingFinalUnclipped_sma_iqr  \
file                                                                             
all_wavs/qstimme_Orschulik_seg_00000000.wav                           0.049390   
all_wavs/qstimme_Orschulik_seg_00000001.wav                           0.038767   
all_wavs/qstimme_Orschulik_seg_00000002.wav                           0.054929   

                                             lld_voicingFinalUnclipped_sma_median  \
file                                                                                
all_wavs/qstimme_Orschulik_seg_00000000.wav                              0.793239   
all_wavs/qstimme_Orschulik_seg_00000001.wav                              0.776797   
all_wavs/qstimme_Orschulik_seg_00000002.wav                              0.781291   

                                             lld_voicingFinalUnclipped_sma_peak  \
file                                                                              
all_wavs/qstimme_Orschulik_seg_00000000.wav                            0.029250   
all_wavs/qstimme_Orschulik_seg_00000001.wav                            0.042688   
all_wavs/qstimme_Orschulik_seg_00000002.wav                            0.038880   

                                             lld_voicingFinalUnclipped_sma_q_kurtosis  \
file                                                                                    
all_wavs/qstimme_Orschulik_seg_00000000.wav                                  1.608404   
all_wavs/qstimme_Orschulik_seg_00000001.wav                                  2.565036   
all_wavs/qstimme_Orschulik_seg_00000002.wav                                  2.802727   

                                             lld_voicingFinalUnclipped_sma_q_range5  \
file                                                                                  
all_wavs/qstimme_Orschulik_seg_00000000.wav                                0.128187   
all_wavs/qstimme_Orschulik_seg_00000001.wav                                0.126647   
all_wavs/qstimme_Orschulik_seg_00000002.wav                                0.201142   

                                             ...  rhy_f0_sm3    rhy_f0_sm4  \
file                                         ...                             
all_wavs/qstimme_Orschulik_seg_00000000.wav  ...   12.722556   5217.711721   
all_wavs/qstimme_Orschulik_seg_00000001.wav  ...   -0.058448   1082.281968   
all_wavs/qstimme_Orschulik_seg_00000002.wav  ...   14.688077  12639.381163   

                                             

### Automatic speech recognition

In [17]:
import asr as speechrec
asr = speechrec.ASR('eng')
df_asr = asr.extract_from_index(mld_df, tmp='./tmp',  margin= 0.2)

  0%|                                                    | 0/13 [00:00<?, ?it/s]

ASR ...


100%|███████████████████████████████████████████| 13/13 [00:09<00:00,  1.37it/s]


In [18]:
df_asr.head()

asr
file                                                                                          
all_wavs/qstimme_Orschulik_seg_00000000.wav                 the world's first genderless voice
all_wavs/qstimme_Orschulik_seg_00000001.wav  think of me like Siri or Alexa but neither mal...
all_wavs/qstimme_Orschulik_seg_00000002.wav  add created for a future will be a no longer d...
all_wavs/qstimme_Orschulik_seg_00000003.wav                            how we define ourselves
all_wavs/qstimme_Orschulik_seg_00000004.wav  my voice was recorded by people who need to id...

In [19]:
import nltk
nltk.download('wordnet')

fex = tf.PsyText(language="eng")
ling_df = pd.DataFrame()
t_all = ''
for t in df_asr.asr:
    t_all = t_all+' '+t
    try:
        feat = fex.extract_from_string(t)
    except ZeroDivisionError:
        feat = pd.DataFrame()
    ling_df = ling_df.append(feat)

[nltk_data] Downloading package wordnet to
[nltk_data]     /home/audeering.local/morschulik/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [20]:
ling_df.head()

N  adjectives   adverbs  assent  conjunctions  content_words  \
0   6    0.166667  0.000000     0.0          0.00       0.666667   
0  12    0.166667  0.000000     0.0          0.25       0.416667   
0  13    0.000000  0.153846     0.0          0.00       0.307692   
0   4    0.000000  0.000000     0.0          0.00       0.250000   
0  13    0.000000  0.000000     0.0          0.00       0.461538   

   cw_typeTokenRatio  function_words  fw_typeTokenRatio  indefinite_pronouns  \
0                1.0        0.333333           1.000000             0.000000   
0                1.0        0.583333           1.000000             0.083333   
0                1.0        0.692308           0.888889             0.000000   
0                1.0        0.750000           1.000000             0.000000   
0                1.0        0.538462           1.000000             0.000000   

   ...  pronouns_1s  pronouns_2  pronouns_3  quantifiers  sentiment  \
0  ...     0.000000         0.0         0.0          0.0   0.333333   
0  ...     0.083333         0.0         0.0          0.0   0.000000   
0  ...     0.000000         0.0         0.0          0.0   0.076923   
0  ...     0.000000         0.0         0.0          0.0   0.250000   
0  ...     0.153846         0.0         0.0          0.0   0.153846   

   sentiment_negative  sentiment_positive  typeTokenRatio  verb_noun_ratio  \
0                 0.0            0.156407        1.000000                0   
0                 0.0            0.000000        1.000000                0   
0                 0.0            0.022948        0.923077                0   
0                 0.0            0.031907        1.000000                0   
0                 0.0            0.036058        1.000000                0   

      verbs  
0  0.000000  
0  0.000000  
0  0.230769  
0  0.250000  
0  0.307692  

[5 rows x 36 columns]

In [21]:
t_all

" the world's first genderless voice think of me like Siri or Alexa but neither male nor female add created for a future will be a no longer defined by g how we define ourselves my voice was recorded by people who need to identify as male Norfolk then altered to sound engineer putting my voice between 145 and 175 arrange to find by audio Research to me to become a third option for voice assistant share my voice with Apple Amazon Google and micro and together we can ensure technology recognisers thanks for this"

In [22]:
feat = fex.extract_from_string(t_all)

In [23]:
feat

N  adjectives   adverbs  assent  conjunctions  content_words  \
0  95    0.073684  0.042105     0.0      0.063158       0.484211   

   cw_typeTokenRatio  function_words  fw_typeTokenRatio  indefinite_pronouns  \
0           0.891304        0.515789           0.653061             0.010526   

   ...  pronouns_1s  pronouns_2  pronouns_3  quantifiers  sentiment  \
0  ...     0.084211         0.0         0.0          0.0   0.157895   

   sentiment_negative  sentiment_positive  typeTokenRatio  verb_noun_ratio  \
0                 0.0             0.05244        0.768421                0   

      verbs  
0  0.147368  

[1 rows x 36 columns]

In [24]:
t_all

" the world's first genderless voice think of me like Siri or Alexa but neither male nor female add created for a future will be a no longer defined by g how we define ourselves my voice was recorded by people who need to identify as male Norfolk then altered to sound engineer putting my voice between 145 and 175 arrange to find by audio Research to me to become a third option for voice assistant share my voice with Apple Amazon Google and micro and together we can ensure technology recognisers thanks for this"